### PR TITLE
Charlesmchen/voice messages ui

### PR DESCRIPTION
**Pull request by charlesmchen**
See: https://github.com/WhisperSystems/SignalServiceKit/pull/212

PTAL @michaelkirk

### CODE REVIEW

File: Signal/src/Models/TSMessageAdapaters/TSVideoAttachmentAdapter.m [hunk 1]
```diff
@@ -259,10 +204,149 @@ - (UIView *)mediaView {
...
+    }
+
+    return imageView;
+}
+
+- (BOOL)isVoiceMessage
+{
+    OWSAssert([self isAudio]);
+
+    return (self.attachment.isVoiceMessage || self.attachment.filename.length < 1);
```
Comment (michaelkirk) [inline Signal/src/Models/TSMessageAdapaters/TSVideoAttachmentAdapter.m]: nit: maybe a comment about the legacy/arcane nature of `self.attachment.filename.length < 1`

File: Signal/src/UIColor+OWS.m [hunk 2]
```diff
@@ -125,4 +126,17 @@ + (UIColor *)colorWithRGBHex:(unsigned long)value
     return [UIColor colorWithRed:red green:green blue:blue alpha:1.f];
 }
 
+- (UIColor *)blendWithColor:(UIColor *)otherColor alpha:(CGFloat)alpha
```
Comment (michaelkirk) [inline Signal/src/UIColor+OWS.m]: nice.

File: Signal/src/views/AudioProgressView.swift [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+    var progress: CGFloat = 0 {
+        didSet {
+            if oldValue != progress {
+                updateContent()
+            }
+        }
+    }
+
+    @available(*, unavailable, message:"use delegate: constructor instead.")
```
Comment (michaelkirk) [inline Signal/src/views/AudioProgressView.swift]: nit: unavailable comment seems like it should be like "use init() constructor instead"

File: Signal/src/Models/TSMessageAdapaters/TSVideoAttachmentAdapter.m (hunk 1 above)
Comment (charlesmchen) [inline Signal/src/Models/TSMessageAdapaters/TSVideoAttachmentAdapter.m]: Done.

File: Signal/src/views/AudioProgressView.swift (hunk 3 above)
Comment (charlesmchen) [inline Signal/src/views/AudioProgressView.swift]: Done.

**Comment by charlesmchen** [comment]
> progress bar of the playing message bounces all around.

Fixed. PTAL @michaelkirk

File: Signal/src/UIColor+OWS.m [hunk 4]
```diff
@@ -129,14 +129,22 @@ + (UIColor *)colorWithRGBHex:(unsigned long)value
...
+        [self getRed:&r0 green:&g0 blue:&b0 alpha:&a0];
+    OWSAssert(result)
+
+        CGFloat r1,
+        g1, b1, a1;
+#ifdef DEBUG
+    result =
+#endif
+        [otherColor getRed:&r1 green:&g1 blue:&b1 alpha:&a1];
+    OWSAssert(result) return [UIColor colorWithRed:CGFloatLerp(r0, r1, alpha)
```
Comment (michaelkirk) [inline Signal/src/UIColor+OWS.m]: nit: weird spacing, missing `;` in your assert macros?

Also is the ifdef there to avoid an unused error?

**Comment by michaelkirk** [comment]
Sorry, I actually meant to LGTM with the last review.

But I'm curious...

> progress bar of the playing message bounces all around.

How did you fix this? I don't see any relevant changes since the last review.

**Comment by charlesmchen** [comment]
> How did you fix this? I don't see any relevant changes since the last review.

We should cancel any audio attachment playback when we start recording a voice message.

```
   // Cancel any ongoing audio playback.
    [self.audioAttachmentPlayer stop];
    self.audioAttachmentPlayer = nil;
```

File: Signal/src/UIColor+OWS.m (hunk 4 above)
Comment (charlesmchen) [inline Signal/src/UIColor+OWS.m]: Exactly.

Comment (charlesmchen) [inline Signal/src/UIColor+OWS.m]: Add I added the semicolons.

**Pull request merged**